java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/CodeLensUtils.java

```diff
@@ -0,0 +1,81 @@
+/*******************************************************************************
+* Copyright (c) 2018 Microsoft Corporation and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License v1.0
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package com.microsoft.java.test.plugin.util;
+
+import com.microsoft.java.test.plugin.model.TestItem;
+import com.microsoft.java.test.plugin.model.TestKind;
+import com.microsoft.java.test.plugin.model.TestLevel;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.handlers.DocumentLifeCycleHandler;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+@SuppressWarnings("restriction")
+public class CodeLensUtils {
```

Comment:
Decouple the underlying concept from the UI for flexibility.  

The CodeLens/TestExplorer are the UI elements. Here, we map the UI elements to the underlying utils, which will make the underlying fragile.  You should provide the API and let the UI decide how to call these APIs. Thus, you can decouple the UI from the underlying system.

Reply:
Follow up:
- Remain the TestSearchUtils there
- Extract out the methods which are used to construct the TestItem into a new file: TestItemsUtils
- Extract out the methods which are used to deal with the test frameworks into a new file: TestFrameworkUtils.